### ISSUE TITLE
[Question] Evaluating dimensionality reduction?

### ISSUE THREAD

**Issue by csanhuezalobos** [lmcinnes/umap#6]
Hello Leland,

Thank you for sharing this new algorithm.
I have a question regarding evaluation measures of dimensionality reduction methods. I'm aware of trustworthiness and continuity, but I'm looking for measures that can handle large datasets.

I found the paper "[Scale-independent quality criteria for dimensionality reduction](https://perso.uclouvain.be/michel.verleysen/papers/patreclet10jl.pdf)" which is an alternative quality measure, but it is still for small datasets.

How are you evaluating umap against other approaches at the moment?

**Comment by lmcinnes** [comment]
Mostly I've been using trustworthiness over varying neighborhood sizes,
which, I agree, is not tractable for large data sets. With enough compute
time (and some judicious pre-compute and code tuning (cython or numba) I
can manage, for example, the full MNIST test set. That means that, for now
at least, I am only really comparing on small datasets. It is notable,
however, that this is true of almost all the literature on manifold based
dimension reduction techniques, so I am at least comparable with the rest
of the literature.

On Tue, Oct 10, 2017 at 10:32 AM, Claudio Sanhueza <notifications@github.com

**Comment by csanhuezalobos** [comment]
Thanks for the hints, Leland. I will try to implement my version of the metrics for large datasets or, at least, a methodology to do it with what is available.

BTW, there is a new MNIST-like dataset called "[fashion MNIST](https://github.com/zalandoresearch/fashion-mnist)" if you want to test (released Aug. 2017). They argue we should move away from MNIST to test new algorithms (see the link for details).

**Comment by lmcinnes** [comment]
I have tried UMAP on fashion MNIST. It does not magically separate the 10
classes (at least not as tidily as it does with digits), but the classes
that it mixes are reasonable under the circumstances. It is, certainly, a
more interesting dataset on which to try such algorithms.

**Comment by csanhuezalobos** [comment]
I got these visualizations of fashion mnist with t-SNE and LargeVis (trains+test=70000 using labels only for the colors). It looks like LargeVis looks better, but then when I evaluate both of them using a kNN classifier (with 10 runs for each training % and using the mean) t-SNE is a better embedding for this task.(_Should I trust my eyes or the numbers?_)

Do you have a visualization using umap that you can share?


![fashion_mnist_tsne](https://user-images.githubusercontent.com/1295756/31478475-5481c1aa-af5c-11e7-8eb8-b33b1ace5824.png)
![fashion_mnist_largevis](https://user-images.githubusercontent.com/1295756/31478481-63151adc-af5c-11e7-84d9-efafcca9bf1c.png)
![k50_fashion_mnist](https://user-images.githubusercontent.com/1295756/31478971-eeb4f122-af5f-11e7-93fa-626cf18ef8af.png)
![k100_fashion_mnist](https://user-images.githubusercontent.com/1295756/31478712-07f2b1da-af5e-11e7-84c5-e1e6ef93c482.png)
![k500_fashion_mnist](https://user-images.githubusercontent.com/1295756/31478947-b065bbea-af5f-11e7-868e-c6f707dc3f70.png)

**Comment by lmcinnes** [comment]
I can dig up a visualization I think. It is closer to LargeVis in appearance, but still a little different. As to what to trust; I think you have to trust both to some extent. t-SNE does do some thing right, and those curves do matter, so despite the clearly better appearance of LargeVis there seems to be something deceptive going on underneath it all.

**Comment by lmcinnes** [comment]
Here's what UMAP did:

![image](https://user-images.githubusercontent.com/11962885/31576785-202ef6e4-b0fa-11e7-9b03-c84ed56c9e92.png)

As I said, more similar to LargeVis. It is worth noting that UMAP has kept some of the groups together where LargeVis split them into multiple blobs (the royal blue category in your LargeVis plot for example). I wonder if that may effect the kNN-classifier accuracy?

**Comment by lmcinnes** [comment]
I tweaked the min_dist parameter (which defines how closely the embedding should pack points together in the embedded space) to compress things less (and hence resemble the t-SNE result more) and got this:

![image](https://user-images.githubusercontent.com/11962885/31577598-929a7adc-b109-11e7-9cb3-ce1c73f1a1b8.png)
Still very similar (up to rotation) but less aggressive in separating clusters and showing a little more of the interconnected structure.  I believe this would almost certainly embed a whole lot better in 3 or 4 dimensions.

**Comment by csanhuezalobos** [comment]
Thank you for sharing, UMAP is doing great (visually). I definitely need to study the details of your implementation. Are you planning to submit a preprint soon? (just trying to decide if I wait for your document or I should jump to implementation instead)

I believe a less aggressive separation would lead to better k-NN classifiers performance, but we should evaluate with trustworthiness and continuity anyway (or others like the scale-independent criteria).

**Comment by lmcinnes** [comment]
I'm struggling to find time to short up all the math and get the preprint done (because I really want sound explanations of *why* things work, which means getting good explanations well hammered out). It will be a little while yet unfortunately. The code may be a little hard to follow, but check the ``numba`` branch as that has code that is, perhaps easier to wrap one's head around. The preprint will probably help rather a lot though. Thanks for the extra reminder that I really need to get to work on getting that done.

**Comment by lionely** [comment]
Is it too much to ask for a code example displaying an implementation of "trustworthiness" and "continuity"? I'm trying to evaluate the quality of dimensionality reductions acquired from t-SNE.

Any help would be greatly appreciated!

**Comment by lmcinnes** [comment]
There's something in https://github.com/lmcinnes/umap/blob/master/umap/validation.py and you can see https://github.com/scikit-learn/scikit-learn/blob/ccd3331f7eb3468ac96222dc5350e58c58ccba20/sklearn/manifold/t_sne.py#L394 for a (semi-canonical) implementation.

**Comment by hoangthienan95** [comment]
Are "Trustworthiness" and "continuity" still the two best measures for evaluating the embedding? In `validation.py`, I see the parameter `max_k` in the function `trustworthiness_vector`. How do I choose this parameter?

Also, kind of related,  you said there are [some guidance on how many `n_components` we should choose](https://github.com/lmcinnes/umap/issues/257#issuecomment-504248922). Any update on that? TIA!

**Comment by cglopezespina** [comment]
Is trustworthiness a good method for selection of UMAP parameters. It is mentioned above, but I have not seen it in any other resource.

**Comment by kbaran1998** [comment]
Based on my experience, I recommend using the [ZADU](https://github.com/hj-n/zadu) package for this task. It is based on the information provided in [this paper](https://arxiv.org/pdf/2308.00282).
